text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+
+By deciding to "cut" the block chain at a specific parent, we can apply a similar garbage collection algorithm:
+
+1. Decide which blocks to keep and add them to the "objects to keep" set (*i.e.* a starting block and all descendents).
+1. For each block, visit all of its dependencies (deploys, finality signatures) and add them to an "objects to keep" set.
+1. Visit each block, deploy and finality signature, remove them if not inside the "objects to keep" set.
+
+In a manner similar to the trie store, we can employ a mark-and-sweep version of this pruning algorithm as well.
+
+This again reduces the problem on selecting which blocks to keep.
+
+### Pruning summary
+
+Both trie-store as well as storage can be trimmed down by garbage collection, with savings in the latter probably being more substantial than the former.
+
+In both cases the problem can reduced to identifying which root hashes or blocks must be kept and which can be discarded. Once this information is available, garbage collection becomes straight forward.
+
+An "archival node" is a node on which garbage collection is either disabled or restricted to non-transitive state information, *i.e.* it keeps all root hashes except those created in the midst of execution.
+
+### Fast-syncing
+
+Fast syncing, while beneficial in its own right, is also of interest from a storage perspective, as being able to recover information quickly from the network or external sources influences our decisions about what information must be kept.
+
+The "fast" part of fast-syncing, as compared to the status quo, is realized from two factors, namely
+
+* **side-loading** data from outside the network through external sources and
+* **transfering** state, namely trie-store state, which is currently not transferrable at all.
+
+### Side-loading
+
+We create a common trait for Blocks, Deploys or any other transferable object that states how it can be serialized and defines a hash function for its serialized representation, as well as a key to separate the hash space. As an example, a block whose hash is `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2` would be adressable as `blockv1_5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`. The `blockv1` denotes a complete scheme, that is it also pins the serialization format. While it is not necessary to separate the hash value space to avoid collisions, the prefix allows for multiple representations, should the format of a serialization or an internal data structure change at some point. We refer to data structures that satisfy this condition as *objects* from here on out.
```

Comment:
I think we should cleanly separate retrieval of a block, deploy, trie node from a peer vs retrieval from outside the network.  There is no need to have p2p message sizes increased by forcing URL-friendly representations of hashes.

Reply:
> I think we should cleanly separate retrieval of a block, deploy, trie node from a peer vs retrieval from outside the network. 

I'm not sure we're not making our lives extra difficult here. By nature of coming from different components and different transports, objects are already separately retrieved, what extra separation do we need?

> There is no need to have p2p message sizes increased by forcing URL-friendly representations of hashes.

I may have phrased this poorly, but there address is just for URL purposes on outside stores - there's no need to send the hash as a hexstring in our own wireformat.